cps/kni.c

```diff
@@ -55,6 +62,240 @@ kni_change_mtu(uint8_t port_id, unsigned new_mtu)
+	 * we have to fork a child process to issue the request
+	 * while the parent calls rte_kni_handle_request().
+	 */
+
+	pid = fork();
+	if (pid == -1) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"cps: fork failed, can't modify KNI %s link: %s\n",
+			kni_name, strerror(errno));
+		return -1;
+	} else if (pid == 0) {
+		/*
+		 * Send request to kernel, which will be sent back
+		 * to userspace for the parent to handle.
+		 */
+		int ret = mnl_socket_sendto(nl, nlh, nlh->nlmsg_len);
+		if (ret < 0) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"cps: mnl_socket_sendto: cannot bring KNI %s %s: %s\n",
+				kni_name, if_up ? "up" : "down",
+				strerror(errno));
+			_exit(EXIT_FAILURE);
+		}
+		_exit(EXIT_SUCCESS);
+	} else {
+		unsigned int attempts = 0;
+		unsigned int portid = mnl_socket_get_portid(nl);
+		do {
+			int status;
+
+			/* Try to process child's request. */
+			int ret = rte_kni_handle_request(kni);
+			if (ret < 0) {
+				RTE_LOG(ERR, KNI, "%s: error in handling userspace request\n",
+					__func__);
+				goto next;
+			}
+
+			/* Check if child has finished submitting request. */
+			ret = waitpid(pid, &status, WNOHANG);
+			if (ret == 0) {
+				/* Keep trying to handle the KNI request. */
+				goto next;
+			} else if (ret == -1) {
+				RTE_LOG(ERR, GATEKEEPER, "cps: waitpid: %s\n",
+					strerror(errno));
+				goto kill;
+			}
+
+			/* Verify interface configuration message was valid. */
+			ret = mnl_socket_recvfrom(nl, buf, sizeof(buf));
+			if (ret == -1) {
+				RTE_LOG(ERR, GATEKEEPER, "cps: mnl_socket_recvfrom: cannot bring KNI %s %s: %s\n",
+					kni_name, if_up ? "up" : "down",
+					strerror(errno));
+				return ret;
+			}
+
+			ret = mnl_cb_run(buf, ret, seq, portid, NULL, NULL);
+			if (ret == -1) {
+				RTE_LOG(ERR, GATEKEEPER, "cps: mnl_cb_run: cannot bring KNI %s %s: %s\n",
+					kni_name, if_up ? "up" : "down",
+					strerror(errno));
+				return ret;
+			}
+
+			return 0;
+next:
+			attempts++;
+			sleep(1);
+		} while (attempts < NUM_ATTEMPTS_KNI_LINK_SET);
+kill:
+		kill(pid, SIGTERM);
+	}
+
+	return ret;
+}
+
+int
+kni_config(struct rte_kni *kni, struct gatekeeper_if *iface)
+{
+	struct mnl_socket *nl;
+	const char *kni_name = rte_kni_get_name(kni);
+	int ret = 0;
+
+	nl = mnl_socket_open(NETLINK_ROUTE);
+	if (nl == NULL) {
+		RTE_LOG(ERR, GATEKEEPER, "cps: mnl_socket_open: %s\n",
+			strerror(errno));
+		return -1;
+	}
+
+	ret = mnl_socket_bind(nl, 0, MNL_SOCKET_AUTOPID);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER, "cps: mnl_socket_bind: %s\n",
+			strerror(errno));
+		goto close;
+	}
+
+	/* Add global and link-local IP addresses. */
+	if (ipv4_if_configured(iface)) {
+		ret = add_ipaddr(nl, AF_INET, &iface->ip4_addr,
+			iface->ip4_addr_plen, kni_name);
+		if (ret < 0)
+			goto close;
+	}
+
+	if (ipv6_if_configured(iface)) {
+		ret = add_ipaddr(nl, AF_INET6, &iface->ip6_addr,
+			iface->ip6_addr_plen, kni_name);
+		if (ret < 0)
+			goto close;
+
+		ret = add_ipaddr(nl, AF_INET6, &iface->ll_ip6_addr,
```

Comment:
Do we really need to add the link local address? According to the following page, the "Kernel automatically assigns a MAC-derived link-local address to any active Ethernet interface."

https://fedoraproject.org/wiki/Networking/Addressing#Link-local_addresses_.28mostly_IPv6-only.29

I didn't read it through, but the code in the kernel is around here:

https://github.com/torvalds/linux/blob/v4.10/net/ipv6/addrconf.c#L2932

Credits: Andre provided me with the information above.